### PR TITLE
nixos/tests/installer: add bpool for grub files in zfsroot test

### PR DESCRIPTION
**Pull request by misuzu**
## Description of changes
Fix nixosTests.zfs.installer
See also https://github.com/NixOS/nixpkgs/pull/262951

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [x] i686-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ElvishJerricco** [comment]
We could also do `zpool create -o compatibility=grub2 ...` instead. It appears that grub's support for ZFS isn't what changed here; the ZFS feature set enabled by default is now beyond what grub supports.

(That said, I generally recommend against using grub's ZFS support in the first place, so I have no issues with adding a `/boot` partition anyway)

**Comment by misuzu** [comment]
> That said, I generally recommend against using grub's ZFS support in the first place

We should probably just get rid of zfs support in grub

**Comment by misuzu** [comment]
> We could also do `zpool create -o compatibility=grub2 ...` instead. It appears that grub's support for ZFS isn't what changed here; the ZFS feature set enabled by default is now beyond what grub supports.

Can confirm that this works

**Comment by ElvishJerricco** [comment]
> We should probably just get rid of zfs support in grub

This probably isn't advisable because the [openzfs-docs](https://openzfs.github.io/openzfs-docs/Getting%20Started/NixOS/Root%20on%20ZFS.html) instructions for root on ZFS still include a ZFS bpool for grub (for all distros; not just NixOS).

Given that, I'm tempted to say we should just use `-o compatibility=grub2`, or otherwise do the bpool style steup in our installer test. Not because I think it's a good idea to set a system up that way, but because there will be people who do. So we probably want to support it.

**Comment by misuzu** [comment]
> or otherwise do the bpool style steup in our installer test

I think this is the best option. Updated the pr accordingly

**Comment by misuzu** [comment]
@ofborg test zfs.installer

**Comment by ElvishJerricco** [comment]
Could just use `boot.zfs.forceImportAll = true;`, but realistically it'd be better to put a `machine.succeed("zpool export -a")` in the `Shutdown system after installation` section of the installer test script after `umount -R /mnt`. Probably means adding a `postInstallCommands` argument or something to that function?

**Comment by misuzu** [comment]
> postInstallCommands

I've added postInstallCommands to makeInstallerTest

**Review by ElvishJerricco — APPROVED** [review]
LGTM

**Comment by misuzu** [comment]
Does work on i686-linux, though it's painfully slow (40min on my machine)

**Review by adamcstephens — APPROVED** [review]
Built successfully on x86_64 and i686.

File: nixos/tests/installer.nix [hunk 1]
```diff
@@ -682,20 +686,32 @@ in {
     createPartitions = ''
       machine.succeed(
           "flock /dev/vda parted --script /dev/vda -- mklabel msdos"
-          + " mkpart primary linux-swap 1M 1024M"
-          + " mkpart primary 1024M -1s",
+          + " mkpart primary 1M 100MB"  # bpool
```
Comment (Mic92) [inline nixos/tests/installer.nix:689]: Looks good for this PR, but maybe we should start considering not testing to boot from zfs at all? I think the grub integration for zfs is not very well maintained and we mention this in no installation guide for zfs on NixOS. Given that most people now use EFI on their regular installations, we might just test this use-case instead?

Comment (Mic92) [reply nixos/tests/installer.nix:689]: Ok. Just read the discussion above. I think we sort of agree, but let's just use your solution for now.

**Comment by Luflosi** [comment]
I'm confused why `-o compatibility=grub2` works. According to the [ZFS documentation](https://openzfs.github.io/openzfs-docs/man/master/7/zpool-features.7.html#Compatibility_feature_sets) the `compatibility` filename is either an absolute path or a relative path to `/etc/zfs/compatibility.d` or `/usr/share/zfs/compatibility.d`. `grub2` is not an absolute path and we don't have anything at `/etc/zfs/compatibility.d` or `/usr/share/zfs/compatibility.d`. I guess there must be some undocumented behaviour that allows this somehow.

**Comment by Luflosi** [comment]
Also see #223412.

**Pull request merged**